adr/ADR-19.md

```diff
@@ -0,0 +1,281 @@
+    // If empty means this is a link to the whole store, like a directory.
+    Name string `json:"name,omitempty"`
+}
+```
+
+### ObjectMetaOptions
+
+```go
+type ObjectMetaOptions struct {
+    Link      *ObjectLink `json:"link,omitempty"`
+    ChunkSize uint32      `json:"max_chunk_size,omitempty"`
+}
+```
+
+### ObjectMeta 
+
+Object Meta is high level information about an object.
+
+```go
+type ObjectMeta struct {
+    Name        string `json:"name"`
+    Description string `json:"description,omitempty"`
+    Headers     Header `json:"headers,omitempty"`
+
+    // Optional options.
+    Opts *ObjectMetaOptions `json:"options,omitempty"`
+}
+```
+
+### ObjectInfo 
+
+Object Info is meta plus instance information.
+
+```go
+type ObjectInfo struct {
+    ObjectMeta
+    
+    Bucket  string    `json:"bucket"`
+    
+    NUID    string    `json:"nuid"`
+    
+    // the total object size in bytes
+    Size    uint64    `json:"size"`
+    
+    ModTime time.Time `json:"mtime"`
+    
+    // the total number of chunks
+    Chunks  uint32    `json:"chunks"`
+    
+    // as in http, <digest-algorithm>=<digest-value>
+    Digest  string    `json:"digest,omitempty"`
+    
+    Deleted bool      `json:"deleted,omitempty"`
+}
+```
+
+### ObjectStoreStatus
+
+The status of an object
+
+```go
+type ObjectStoreStatus interface {
+    // Bucket is the name of the bucket
+    Bucket() string
+    
+    // Description is the description supplied when creating the bucket
+    Description() string
+    
+    // TTL indicates how long objects are kept in the bucket
+    TTL() time.Duration
+    
+    // Storage indicates the underlying JetStream storage technology used to store data
+    Storage() StorageType
+    
+    // Replicas indicates how many storage replicas are kept for the data in the bucket
+    Replicas() int
+    
+    // Sealed indicates the stream is sealed and cannot be modified in any way
+    Sealed() bool
+    
+    // Size is the combined size of all data in the bucket including metadata, in bytes
+    Size() uint64
+    
+    // BackingStore provides details about the underlying storage
+    BackingStore() string
+}    
+```
+
+## Functional Interfaces
+
+### ObjectStoreManager 
+
+Object Store manger creates, loads and deletes Object Stores
+
+```go
+type ObjectStoreManager interface {
+    // ObjectStore will lookup and bind to an existing object store instance.
+    ObjectStore(bucket string) (ObjectStore, error)
+    
+    // CreateObjectStore will create an object store.
+    CreateObjectStore(cfg *ObjectStoreConfig) (ObjectStore, error)
+    
+    // DeleteObjectStore will delete the underlying stream for the named object.
+    DeleteObjectStore(bucket string) error
+}
+```
+
+### ObjectStore 
+
+Storing large objects efficiently
+
+```go
+type ObjectStore interface {
+    // Put will place the contents from the reader into a new object.
+    Put(obj *ObjectMeta, reader io.Reader, opts ...ObjectOpt) (*ObjectInfo, error)
```

Comment:
I think like KV we should specify the core API and leave convenience functions up to specific languages, so the `GetBytes` and `GetString` variants can be left as a thing where languages do what works for them rather than specify here which implies its needed

Reply:
Ok I'll add notes in the comment to say optional or maybe separate out an optional section.

Reply:
I am still seeing the variants RI is describing...

Reply:
I had added this "Please note that anything that is commented as a "convenience function" is recommended but optional if it does not make sense for the client language."